src/awkward/_nplikes/virtual.py

```diff
@@ -0,0 +1,301 @@
+from awkward._nplikes.numpy_like import NumpyLike, NumpyMetadata
+from awkward._nplikes.shape import ShapeItem, unknown_length
+from awkward._operators import NDArrayOperatorsMixin
+from awkward._regularize import is_integer
+from awkward._typing import TYPE_CHECKING, Any, Callable, DType, Self
+from awkward._util import Sentinel
+
+np = NumpyMetadata.instance()
+
+if TYPE_CHECKING:
+    from numpy.typing import DTypeLike
+
+
+UNMATERIALIZED = Sentinel("<UNMATERIALIZED>", None)
+
+
+def materialize_if_virtual(*args: Any) -> tuple[Any, ...]:
+    """
+    A little helper function to materialize all virtual arrays in a list of arrays.
+    """
+    return tuple(
+        arg.materialize() if isinstance(arg, VirtualArray) else arg for arg in args
+    )
+
+
+class VirtualArray(NDArrayOperatorsMixin, ArrayLike):
+    """
+    Implements a virtual array to be used as a buffer inside layouts.
+    Virtual arrays are tied to specific nplikes and only numpy and cupy nplikes are allowed.
+    Therefore, virtual arrays are only allowed to generate `numpy.ndarray`s or `cupy.ndarray`s when materialized.
+    The arrays are generated via a generator function that is passed to the constructor.
+    All virtual arrays also required to have a known dtype and shape and `unknown_length` is not currently allowed in the shape.
+    Some operations (such as trivial slicing) maintain virtualness and return a new virtual array.
+    Others are required to access the underlying data of the array and therefore materialize it.
+    The materialized arrays are cached on themselves in the `_array` property.
+    Subsequent virtual arrays that originate from some virtual array will hit the cache of their parents if there is any.
+    """
+
+    def __init__(
+        self,
+        nplike: NumpyLike,
+        shape: tuple[ShapeItem, ...],
+        dtype: DTypeLike,
+        generator: Callable[[], ArrayLike],
+    ) -> None:
+        if not nplike.supports_virtual_arrays:
+            raise TypeError(
+                f"Only numpy and cupy nplikes are supported for {type(self).__name__}. Received {type(nplike)}"
+            )
+        if any(not is_integer(dim) for dim in shape):
+            raise TypeError(
+                f"Only shapes of integer dimensions are supported for {type(self).__name__}. Received shape {shape}."
+            )
+
+        # array metadata
+        self._nplike = nplike
+        self._shape = shape
+        self._dtype = np.dtype(dtype)
+        self._array: Sentinel | ArrayLike = UNMATERIALIZED
+        self._generator = generator
+
+    @property
+    def dtype(self) -> DType:
+        return self._dtype
+
+    @property
+    def shape(self) -> tuple[ShapeItem, ...]:
+        return self._shape
+
+    @property
+    def ndim(self) -> int:
+        return len(self._shape)
+
+    @property
+    def size(self) -> ShapeItem:
+        if len(self._shape) == 0:
+            return 1
+        return reduce(mul, self._shape)
```

Comment:
Style preference I'd probably prefer
```suggestion
        size: ShapeItem = 1
        for item in self._Shape:
            size *= item
        return size
```